Set TMPDIR to /var/tmp (FP <1.11.1) or /tmp (FP >=1.11.1) to ensure we are within sun_path limit

rany2 opened the pull request:
If TMPDIR is set to something that's ~61 characters long, Chromium
will crash. Below is the stacktrace:

```
0  0x000055555d22591e in base::ImmediateCrash () at ../../base/immediate_crash.h:176
1  logging::CheckFailure () at ../../base/check.h:211
2  (anonymous namespace)::SetupSocket (path=..., sock=0x1b2000254ca8, addr=0x7fffffffc9f0, socklen=<optimized out>) at ../../chrome/browser/process_singleton_posix.cc:302
3  ProcessSingleton::Create (this=0x1b2000254c30) at ../../chrome/browser/process_singleton_posix.cc:1068
4  ProcessSingleton::NotifyOtherProcessWithTimeoutOrCreate (this=0x1b2000254c30, command_line=..., retry_attempts=20, timeout=...) at ../../chrome/browser/process_singleton_posix.cc:967
5  0x0000555558067489 in ProcessSingleton::NotifyOtherProcessOrCreate (this=0x14) at ../../chrome/browser/process_singleton_posix.cc:943
6  ChromeProcessSingleton::NotifyOtherProcessOrCreate (this=0x1b2000254c00) at ../../chrome/browser/chrome_process_singleton.cc:29
7  (anonymous namespace)::AcquireProcessSingleton (user_data_dir=...) at ../../chrome/app/chrome_main_delegate.cc:567
8  ChromeMainDelegate::PostEarlyInitialization (this=0x7fffffffd270, invoked_in=...) at ../../chrome/app/chrome_main_delegate.cc:842
9  0x000055555c622d4f in content::ContentMainRunnerImpl::RunBrowser (this=0x1b200025c000, main_params=..., start_minimal_browser=<optimized out>)
    at ../../content/app/content_main_runner_impl.cc:1209
10 content::ContentMainRunnerImpl::Run (this=0x1b200025c000) at ../../content/app/content_main_runner_impl.cc:1144
11 0x000055555c61fa06 in content::RunContentProcess (content_main_runner=0x1b200025c000, params=...) at ../../content/app/content_main.cc:335
12 content::ContentMain (params=...) at ../../content/app/content_main.cc:348
13 0x0000555558066cf3 in ChromeMain (argc=<optimized out>, argv=<optimized out>) at ../../chrome/app/chrome_main.cc:192
14 0x00007ffff643b08a in __libc_start_call_main () at /usr/lib/x86_64-linux-gnu/libc.so.6
15 0x00007ffff643b14b in __libc_start_main () at /usr/lib/x86_64-linux-gnu/libc.so.6
16 0x0000555557b48025 in _start () at ../sysdeps/x86_64/start.S:115
```

This is because sun_path could be max 108 bytes including \0
and some app IDs might cause this limit to be exceeded.

We set /tmp (ramdisk) if FP version is >=1.11.1 or /var/tmp
(persistent) if FP version is <1.11.1. We don't care
about the persistance aspect, only that the TMPDIR is shared
between all instances of an app ID. Reference for when /tmp
became shared between all instances is available here:
https://github.com/flatpak/flatpak/commit/b65b3f6eadd51bd6600df2c0d07f902a552163d2

This issue is not theoretical and actually blocked:
https://github.com/flathub/flathub/pull/5098

refi64 commented:
Hmm I think this has a con which is that `/var/tmp` is persistent, unlike `$XDG_RUNTIME_DIR`...this is admittedly how it was in Chromium a long time ago (we used `$XDG_CACHE_HOME`) but it's probably not ideal. The only other path we can use is `/tmp` which is [shared these days](https://github.com/flatpak/flatpak/commit/b65b3f6eadd51bd6600df2c0d07f902a552163d2) but only in Flatpak >= 1.11.

The ideal thing to do might be to check `flatpak-version` and use `/tmp` if it's known to be shared, falling back to `/var/tmp` otherwise? I can look into doing it if you don't want to, though we do already have logic to read `/.flatpak-info` in `cobalt-host.c` (the actual hard part would be the version comparison; glibc has `strverscmp`, could also just `sscanf` and compare the numbers by hand, not sure what the best option is).

rany2 commented:
I think I could do this myself, I'll let you know if I need any help @refi64

rany2 committed:
Set TMPDIR to /var/tmp (FP <1.11.1) or /tmp (FP >=1.11.1) to ensure we are within sun_path limit

If TMPDIR is set to something that's ~61 characters long, Chromium
will crash. Below is the stacktrace:

```
0  0x000055555d22591e in base::ImmediateCrash () at ../../base/immediate_crash.h:176
1  logging::CheckFailure () at ../../base/check.h:211
2  (anonymous namespace)::SetupSocket (path=..., sock=0x1b2000254ca8, addr=0x7fffffffc9f0, socklen=<optimized out>) at ../../chrome/browser/process_singleton_posix.cc:302
3  ProcessSingleton::Create (this=0x1b2000254c30) at ../../chrome/browser/process_singleton_posix.cc:1068
4  ProcessSingleton::NotifyOtherProcessWithTimeoutOrCreate (this=0x1b2000254c30, command_line=..., retry_attempts=20, timeout=...) at ../../chrome/browser/process_singleton_posix.cc:967
5  0x0000555558067489 in ProcessSingleton::NotifyOtherProcessOrCreate (this=0x14) at ../../chrome/browser/process_singleton_posix.cc:943
6  ChromeProcessSingleton::NotifyOtherProcessOrCreate (this=0x1b2000254c00) at ../../chrome/browser/chrome_process_singleton.cc:29
7  (anonymous namespace)::AcquireProcessSingleton (user_data_dir=...) at ../../chrome/app/chrome_main_delegate.cc:567
8  ChromeMainDelegate::PostEarlyInitialization (this=0x7fffffffd270, invoked_in=...) at ../../chrome/app/chrome_main_delegate.cc:842
9  0x000055555c622d4f in content::ContentMainRunnerImpl::RunBrowser (this=0x1b200025c000, main_params=..., start_minimal_browser=<optimized out>)
    at ../../content/app/content_main_runner_impl.cc:1209
10 content::ContentMainRunnerImpl::Run (this=0x1b200025c000) at ../../content/app/content_main_runner_impl.cc:1144
11 0x000055555c61fa06 in content::RunContentProcess (content_main_runner=0x1b200025c000, params=...) at ../../content/app/content_main.cc:335
12 content::ContentMain (params=...) at ../../content/app/content_main.cc:348
13 0x0000555558066cf3 in ChromeMain (argc=<optimized out>, argv=<optimized out>) at ../../chrome/app/chrome_main.cc:192
14 0x00007ffff643b08a in __libc_start_call_main () at /usr/lib/x86_64-linux-gnu/libc.so.6
15 0x00007ffff643b14b in __libc_start_main () at /usr/lib/x86_64-linux-gnu/libc.so.6
16 0x0000555557b48025 in _start () at ../sysdeps/x86_64/start.S:115
```

This is because sun_path could be max 108 bytes including \0
and some app IDs might cause this limit to be exceeded.

We set /tmp (ramdisk) if FP version is >=1.11.1 or /var/tmp
(persistent) if FP version is <1.11.1. We don't care
about the persistance aspect, only that the TMPDIR is shared
between all instances of an app ID. Reference for when /tmp
became shared between all instances is available here:
https://github.com/flatpak/flatpak/commit/b65b3f6eadd51bd6600df2c0d07f902a552163d2

This issue is not theoretical and actually blocked:
https://github.com/flathub/flathub/pull/5098

Signed-off-by: rany <ranygh@riseup.net>

rany2 commented:
@refi64 Please let me know if the current solution is acceptable to you, thank you

refi64 commented on src/cobalt-host.c:
```diff
@@ -85,6 +97,51 @@ const char *cobalt_host_get_app_exec(CobaltHost *host, GError **error) {
+    }
+
+    host->fp_version = g_new0(SemVer, 1);
+    g_autoptr(GRegex) regex = g_regex_new("^([0-9]+)\\.([0-9]+)\\.([0-9]+)", 0, 0, error);
```
Bit of a nit: this can just be:

```suggestion
    g_autoptr(GRegex) regex = g_regex_new("^(\\d+)\\.(\\d+)\\.(\\d+)", 0, 0, error);
```

rany2 replied:
I just don't like how confusing those `\\d` look

refi64 commented on src/cobalt-host.h:
```diff
@@ -15,11 +16,15 @@ const char *cobalt_host_get_app_id(CobaltHost *host, GError **error);
 
 const char *cobalt_host_get_app_exec(CobaltHost *host, GError **error);
 
+SemVer *cobalt_host_get_fp_version(CobaltHost *host, GError **error);
```
Also a bit of a nit: this can be private to cobalt-host (at least for now), like `check_for_binary` and such are.

Erick555 commented:
Note that unlike `/run/user/<uid>/<appid>` both `/var/tmp` and `/tmp` in flatpak are different paths than in host which may make sharing files between apps outside sandbox and flatpak difficult. That was problem for mozilla apps which saved temporary downloaded files in TMPDIR I don't know if chromium could face same issue.

> Hmm I think this has a con which is that /var/tmp is persistent, unlike $XDG_RUNTIME_DIR...this is admittedly how it was in Chromium a long time ago (we used $XDG_CACHE_HOME) but it's probably not ideal.

Are you sure about that? I see $XDG_RUNTIME_DIR used in [first](https://github.com/flathub/org.chromium.Chromium/blob/2d95c9333ab40ed5368472193f14b0d82dff40a7/chromium.sh#L29) commit. I don't think chromium ever set persistent TMPDIR.

rany2 commented:
> > Hmm I think this has a con which is that /var/tmp is persistent, unlike $XDG_RUNTIME_DIR...this is admittedly how it was in Chromium a long time ago (we used $XDG_CACHE_HOME) but it's probably not ideal.
> 
> Are you sure about that? I see $XDG_RUNTIME_DIR used in [first](https://github.com/flathub/org.chromium.Chromium/blob/2d95c9333ab40ed5368472193f14b0d82dff40a7/chromium.sh#L29) commit.

I'm not sure if I misunderstood his statement, but the way I understood what he meant could be demonstrated by the following:

```
➜  ~ flatpak run --command=bash com.google.Chrome 
[📦 com.google.Chrome ~]$ touch /var/tmp/hello /tmp/hello
[📦 com.google.Chrome ~]$ ls /var/tmp/hello /tmp/hello
/tmp/hello  /var/tmp/hello
[📦 com.google.Chrome ~]$ 
exit
➜  ~ flatpak run --command=bash com.google.Chrome 
[📦 com.google.Chrome ~]$ ls /var/tmp/hello /tmp/hello
ls: cannot access '/tmp/hello': No such file or directory
/var/tmp/hello
[📦 com.google.Chrome ~]$ 
exit
```

Edit: I don't think XDG_CACHE/RUNTIME_DIR distinction matters in this context if that's what you meant.

rany2 commented:
> Note that unlike `/run/user/<uid>/<appid>` both `/var/tmp` and `/tmp` in flatpak are different paths than in host which may make sharing files between apps outside sandbox and flatpak difficult. That was problem for mozilla apps which saved temporary downloaded files in TMPDIR I don't know if chromium could face same issue.

I think this is worth keeping in mind, but so far I haven't faced any issues. I obviously can't check every instance of when TMPDIR is used by Chromium source, but so far the folder only contains SingletonCookie/Socket which is used for the "check if another  instance is running"  check.

rany2 commented:
@Erick555 I have an ugly alternative if setting TMPDIR to `{,/var}/tmp` causes issues. Essentially, TMPDIR gets set to something like `TMPDIR=/tmp/shared` with `/tmp/shared` being a symlink to `/run/user/<uid>/<appid>`

Edit: i don't think that would work, discard...

Erick555 commented:
> I'm not sure if I misunderstood his statement, but the way I understood what he meant could be demonstrated by the following:

I was referring to:

> this is admittedly how it was in Chromium a long time ago (**we used $XDG_CACHE_HOME**) but it's probably not ideal.

I can't find any trace of chromium used $XDG_CACHE_HOME or anything else except $XDG_RUNTIME_DIR. So persistent TMPDIR  will be brand new thing

rany2 commented:
> So persistent TMPDIR will be brand new thing

Not really, `$XDG_RUNTIME_DIR/app/$FLATPAK_ID` seems to be persistent just not on disk.

```
➜  ~ flatpak run --command=bash com.google.Chrome 
[📦 com.google.Chrome ~]$ touch /var/tmp/test /tmp/test "$XDG_RUNTIME_DIR/app/$FLATPAK_ID/test"
[📦 com.google.Chrome ~]$ ls /var/tmp/test /tmp/test "$XDG_RUNTIME_DIR/app/$FLATPAK_ID/test"
/run/user/1000/app/com.google.Chrome/test  /tmp/test  /var/tmp/test
[📦 com.google.Chrome ~]$ 
exit
➜  ~ flatpak run --command=bash com.google.Chrome 
[📦 com.google.Chrome ~]$ ls /var/tmp/test /tmp/test "$XDG_RUNTIME_DIR/app/$FLATPAK_ID/test"
ls: cannot access '/tmp/test': No such file or directory
/run/user/1000/app/com.google.Chrome/test  /var/tmp/test
[📦 com.google.Chrome ~]$ 
```

Erick555 commented:
Persistent here means written on disk (instead of memory). It can survive system reboot.

refi64 commented:
> Are you sure about that? I see $XDG_RUNTIME_DIR used in [first](https://github.com/flathub/org.chromium.Chromium/blob/2d95c9333ab40ed5368472193f14b0d82dff40a7/chromium.sh#L29) commit.

Oops it was Chrome, not Chromium: https://github.com/flathub/com.google.Chrome/commit/3b7d160b8bf3178166c1565bf04e48f7c37d7b88 You're the one who mentioned it could be changed to `$XDG_RUNTIME_DIR` :sweat_smile: 

I don't *think* surviving reboot is a major issue here other than just causing clutter? Because the state of the socket would be the same as if Chrome just died, which it already can do anyway and it thus has to handle that. That being said, the files accumulating over time could maybe be an issue; I dunno if there's a particularly clean way to solve that?

Some other ideas:

- Maybe the code in cobalt fall back to the current `$XDG_RUNTIME_DIR`-based location and then have the ungoogled flatpak require 1.11.1 at minimum?
- The above, but Cobalt could just check if the path length is greater than `sizeof(sun_path)` itself and only then fall back to `/var/tmp`, basically the worst-case scenario, and it could show a warning dialog about this (we already have some logic for it). But it's not clear how actionable the dialog would be to users, which is already an issue w/ the current dialogs we have here.

rany2 commented:
This PR is getting more complicated than I had anticipated, but I do think that this might be unnecessary as I haven't had any issues with the TMPDIR not being accessible to host.

Erick555 commented:
Ok, then maybe keep current PR as is. The new /tmp may fit well and /var/tmp fallback will die sooner or later when distros update flatpak.

refi64 commented:
Thanks!
